File: src/host/model/control_state.py
from random import randint

from pyplib.xml_parser import parse_int, parse_bool
from pyplib.errors import PP_Game_Action_Error

class phase:
	draw = 0
	pre = 1
	main = 2
	post = 3

	first = draw
	last = post
	all_phases = [draw,pre,main,post]

class super_phase:
	setup = 0
	main = 1
	end = 2

class Control_State:
	def __init__(self,**kwargs):
		if len(kwargs) == 0:
			self.super_phase = super_phase.setup
			self.phase = phase.first
			self.turn_owner = self.decide_first_player()
			self.has_drawn = False
		elif kwargs.has_key('element'):
			element = kwargs['element']
			self.super_phase = parse_int(element,'super_phase')
			self.phase = parse_int(element,'phase')
			self.turn_owner = parse_int(element,'turn_owner')
			self.has_drawn = parse_bool(element,'has_drawn')
		else:
			raise PP_Game_Action_Error("Control State started with an invalid kwarg")

	def xml_output(self,me_uid,me_index,them_uid,them_index,full):
		if not full:
			if self.turn_owner == me_index:
				uid = me_uid
			elif self.turn_owner == them_index:
				uid = them_uid
			else:
				uid = 0
		else:
			uid = self.turn_owner
		xml = '<super_phase>%s</super_phase>'%str(self.super_phase)
		xml += '<phase>%s</phase>'%str(self.phase)
		xml += '<turn_owner>%s</turn_owner>'%str(uid)
		xml += '<has_drawn>%s</has_drawn>'%str(self.has_drawn)
		return xml

	def step_phase(self):
		self.phase += 1
		if self.phase > phase.last:
			self.phase = phase.last
			raise PP_Game_Action_Error("Phase went past the last one")
	def toggle_turn(self,num_players):
		if self.phase == phase.post:
			self.turn_owner += 1
			if self.turn_owner >= num_players: 
				self.turn_owner = 0
			self.phase = phase.draw
			self.has_drawn = False
		else:
			raise PP_Game_Action_Error("Can only end your turn during the post phase")
	def exit_setup_phase(self):
		if self.super_phase == super_phase.setup:
			self.super_phase = super_phase.main
		else:
			raise PP_Game_Action_Error("Can only be performed in setup super phase")

	def is_given_phase(self,given_phase):
		return self.phase==given_phase

	def decide_first_player(self):
		who = randint(0,1)
		who = 0
		return who

File: src/ppbackend/model/mainStruct/ControlState.java
package ppbackend.model.mainStruct;

import java.util.*;
import org.w3c.dom.Element;

import pplib.XmlParser;
import pplib.exceptions.*;

public class ControlState{
	int[] uids;
	int phase;
	int superPhase;
	int turnOwner;
	boolean hasDrawn;

	public ControlState(Set<Integer> uids){
		this.uids = new int[uids.size()];
		int i=0;
		for(Integer integer : uids){
			this.uids[i++] = integer;
		}
		this.phase = Phase.first;
		this.superPhase = SuperPhase.first;
		this.hasDrawn = false;
		this.turnOwner = decideFirstPlayer();
	}

	public ControlState(XmlParser xmlParser, Element element, Set<Integer> uids) throws PPXmlException{
		this.uids = new int[uids.size()];
		int i=0;
		for(Integer integer : uids){
			this.uids[i] = integer;
			i++;
		}
		this.phase = xmlParser.parseInt(element,"phase");
		this.superPhase = xmlParser.parseInt(element,"super_phase");
		this.hasDrawn = xmlParser.parseBool(element,"has_drawn");
		this.turnOwner = xmlParser.parseInt(element,"turn_owner");
	}

	public String xmlOutput(boolean full) throws PPGameActionException{
		String xml = "<super_phase>"+Integer.toString(this.superPhase)+"</super_phase>";
		xml += "<phase>"+Integer.toString(this.phase)+"</phase>";
		xml += "<has_drawn>"+Boolean.toString(this.hasDrawn)+"</has_drawn>";
		xml += "<turn_owner>"+Integer.toString(this.turnOwner)+"</turn_owner>";
		return xml;
	}

	private void canDraw(int uid) throws PPGameActionException{
		this.verifyGivenSuperPhase(SuperPhase.main);
		this.verifyCurrentTurnOwner(uid);
		this.verifyGivenPhase(Phase.draw);
		this.verifyHasNotDrawn();
	}

	public void draw(int uid) throws PPGameActionException{
		this.canDraw(uid);
		this.hasDrawn = true;
	}

	private void canStepPhase(int uid) throws PPGameActionException{
		this.verifyGivenSuperPhase(SuperPhase.main);
		this.verifyCurrentTurnOwner(uid);
	}

	public void stepPhase(int uid) throws PPGameActionException{
		this.canStepPhase(uid);
		this.phase++;
		if(this.phase > Phase.last){
			this.phase = Phase.last;
			throw new PPGameActionException("Phase went past the last one");
		}
	}

	private void canToggleTurn(int uid) throws PPGameActionException{
		this.verifyGivenSuperPhase(SuperPhase.main);
		this.verifyCurrentTurnOwner(uid);
		if(! (this.phase == Phase.post)){
			throw new PPGameActionException("Can only end your turn during the post phase");
		}
	}

	public void toggleTurn(int uid) throws PPGameActionException{
		this.canToggleTurn(uid);
		if(this.uids[0] == uid){
			this.turnOwner = uids[1];
		}
		else{
			this.turnOwner = uids[0];
		}
		this.phase = Phase.draw;
		this.hasDrawn = false;
	}

	private void canPlay(int uid, int[] phases) throws PPGameActionException{
		this.verifyGivenSuperPhase(SuperPhase.main);
		this.verifyCurrentTurnOwner(uid);
		this.verifyGivenPhases(phases);
	}

	public void play(int uid, int[] phases) throws PPGameActionException{
		this.canPlay(uid,phases);
	}

	public void exitSetupSuperPhase() throws PPGameActionException{
		this.verifyGivenSuperPhase(SuperPhase.setup);
		this.superPhase = SuperPhase.main;
	}

	public void exitMainSuperPhase() throws PPGameActionException{
		this.verifyGivenSuperPhase(SuperPhase.main);
		this.superPhase = SuperPhase.end;
	}

	private void verifyGivenSuperPhase(int superPhase) throws PPGameActionException{
		if(!(this.superPhase == superPhase)){
			throw new PPGameActionException("Wrong super phase");
		}
	}

	private void verifyGivenPhase(int givenPhase) throws PPGameActionException{
		if(!(this.phase == givenPhase)){
			throw new PPGameActionException("Wrong phase");
		}
	}

	private void verifyGivenPhases(int[] givenPhases) throws PPGameActionException{
		boolean contains = false;
		for(int phase : givenPhases){
			if(phase == this.phase){
				contains = true;
			}
		}
		if(!contains){
			throw new PPGameActionException("Wrong phase");
		}
	}

	private void verifyHasNotDrawn() throws PPGameActionException{
		if(this.hasDrawn){
			throw new PPGameActionException("Already drew this turn");
		}
	}

	private void verifyCurrentTurnOwner(int uid) throws PPGameActionException{
		if(!(this.turnOwner == uid)){
			throw new PPGameActionException("It is not player "+Integer.toString(uid)+"'s turn");
		}
	}

	private int decideFirstPlayer(){
		int who = this.uids[0];
		return who;
	}

}
